lib/vsdata.cpp

```diff
@@ -47,6 +47,60 @@ void VisualizationSceneScalarData::FixValueRange()
    }
 }
 
+int VisualizationSceneScalarData::GetFunctionAutoRefineFactor(GridFunction &gf)
+{
+   Mesh *mesh = gf.FESpace()->GetMesh();
+   const int order = gf.FESpace()->GetMaxElementOrder();
+
+   // check for integral elements
+   const int dim = mesh->Dimension();
+   const FiniteElementCollection *fec = gf.FESpace()->FEColl();
+   if (fec && fec->GetMapType(dim) == FiniteElement::INTEGRAL)
+   {
+      cout << "Warning: integral elements are non-polynomial in the physical space,\n"
+           << "         consider increasing the refinement by the key 'o'."
+           << endl;
+   }
+
+   return std::max(order, 1);
+}
+
+int VisualizationSceneScalarData::GetAutoRefineFactor()
+{
+   const int dim = mesh->Dimension();
+   const int ne = (dim == 3)?(mesh->GetNBE()):(mesh->GetNE());
+
+   // determine the refinement based on the order of the mesh and grid function
+   int order_ref = GetFunctionAutoRefineFactor();
+
+   // mesh
+   const FiniteElementSpace *nfes = mesh->GetNodalFESpace();
+   if (nfes)
+   {
+      const int order = nfes->GetMaxElementOrder();
+      order_ref = std::max(order_ref, order);
+   }
+
+   // limit the total number of elements
+   int auto_ref_surf_elem = ne*(order_ref+1)*(order_ref+1);
```

Comment:
Dammit, you are right, it is shifted by one as the loop below has `<=` 😬 .

Reply:
Thanks for the catch 😅 , it is fixed in https://github.com/GLVis/glvis/pull/285/commits/73a01745957e62fa8d66c170044c7c5698c68d5b .

Reply:
The formula was right, the name of the variable is wrong -- ultimately we determine `ref` so that the number of vertices in the refined representation is less than or equal to a given max value. So I think you should revert the formula and change the name of the variable to `auto_ref_vert_max` or something like that.

Reply:
The formula was wrong, but rather from different perspective than @justinlaughlin mentioned. The point of the this and following lines is to set `ref=order_ref` when possible, but when the loop below gets to the point `ref==order_ref` then `ne*(ref+1)*(ref+1) <= auto_ref_surf_elem` with `auto_ref_surf_elem = ne*(order_ref+1)*(order_ref+1)`, so `ref` is increased and the algorithm ends with `ref==order_ref+1`.

Reply:
Okay, I understand now. I think your fix makes sense.

Reply:
Ok, I rewrote the algorithm to (DG) vertices and rebalanced the test cases, so how about now @v-dobrev ?